### PR TITLE
fix(factory): reuse canonical implementation, review, and QA workflows

### LINKED ISSUE #561: Reuse existing extension workflows in isolated software-factory runs
### Desired Behavior

Use the existing extensions issue-to-PR, GitHub PR reviewer, and QA Changes behavior in the Docker software factory. Replace the initial duplicated developer/reviewer prompts and raw JSON comments. Adapt only the execution lifecycle and scoped GitHub transport needed for an already provisioned, ephemeral Docker conversation. Keep independent test evidence and guarded merging separate from review presentation.

### Acceptance Criteria

- [ ] Load existing extension implementation/review/QA instructions from their canonical files, without maintaining duplicate task prompts or a custom review formatter.
- [ ] Execute review synchronously in its provisioned Docker conversation and publish the existing native GitHub review with actionable inline findings.
- [ ] Keep the host GitHub credential outside workers; expose only the repository and role operations required by the existing workflows.
- [ ] Verify that the exact current head received a newly posted review and functional QA report before granting acceptance; keep test logs in artifacts.
- [ ] Developer revisions consume actual PR review bodies and inline comments, and preserved work can still be published after a bounded stop.
- [ ] Add focused regression tests for the adapter and permission boundaries, and demonstrate the workflow on the live factory.

This issue is ready-for-dev. Depends on #557 and #559. The prior proposal for a custom formatter and rewriting historical reviews is superseded by reuse of the existing extensions.

### PR DESCRIPTION
**Pull request by neubig**
- [ ] A human has tested these changes.

---

## Why

The initial factory recipe duplicated the existing implementation/review workflows and posted raw acceptance JSON as reviews. It also failed to read native review bodies and inline findings when revising a PR.

## Summary

- Compose the canonical issue-to-PR, GitHub PR Reviewer, and QA Changes workflow sources; package their exact source bytes and hashes with one workspace-independent bundle.
- Supply a repository-scoped `gh api` transport and retain deterministic test execution and guarded publication in the coordinator. The host GitHub credential remains outside workers.
- Require newly published, commit-pinned native review and QA reports with unambiguous passing verdicts; consume reviews and inline feedback during revisions. Keep complete test evidence in workspace artifacts.

## Issue Number

Closes #561

## How to Test

`python -m pytest tests/test_factory_extension_workflows.py tests/test_github_software_factory.py tests/test_github_factory_gateway.py -q` — 46 passed.

Ruff format and checks passed for the changed Python files. Live factory validation is in progress. Identical-bundle local/Docker execution was demonstrated in OpenHands/automation#451; this remains a draft.

## Notes

Review after #559 and #557 (native stack #560). The workflow layer does not inspect workspace kind. OpenHands/automation#450 tracks the shared dispatcher contract required to supply the same conversation environment in both modes. Existing standalone automation scheduling is unchanged; this recipe reuses their workflow definitions with coordinator-owned scheduling and publication.

Runtime communication now uses the public SDK client from OpenHands/software-agent-sdk#5010; the worker Python needs that SDK build until its release. The same entrypoint and bundle work in local and Docker environments. The 46 focused tests still pass after migration.

Live canonical review verification: the existing PR Reviewer workflow posted a native Markdown review with three inline findings on the target application PR. The developer consumed those findings and published a revised head. Machine-readable acceptance artifacts stayed in the workspace. Independent acceptance of that revision is still pending.

Clean source installation now pins SDK #5010 by immutable commit for the test group. `uv run --group test pytest tests/`: **853 passed, 26 skipped**. Replace the integration pin with the SDK release before merging.

### Current validation

Latest fixes supply gateway grants through profile secret references (SDK #5015), validate publication paths, skip closed source issues, make coordinator publication explicit in the canonical builder, preserve pagination filters, and report review failures visibly. The full suite passed 876 tests with 26 skips plus one stale generated bundle check; that check passed after regenerating bundle-index.js. The SDK test pin remains a release dependency before merge.

### CODE REVIEW

File: skills/github-software-factory/scripts/extension_workflows.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        f"Include `<!-- factory-run:{run_id}:{stage} -->` in the review body "
+        "so the coordinator can identify this run's report. "
+        "Preserve the workflow's normal readable report and verdict. "
+        "Never paste a JSON artifact or full test log as the review body.\n"
+    )
+
+
+def implementation_prompt(repository, issue, branch, base_sha, workspace, feedback):
+    workflow = module("skills/github-issue-to-pr/scripts/main.py")
+    prompt = workflow._build_implementation_prompt(
```
[the remark this answers is not in the document]
Comment (neubig) [reply skills/github-software-factory/scripts/extension_workflows.py:58]: Documented the prompt-builder coupling as an explicit integration contract in the README. Canonical helpers and factory ship from one extensions revision, with actual-function integration tests and exact bundle sources. The implementation builder now explicitly supports coordinator-owned publication.

File: skills/github-software-factory/scripts/main.py [hunk 2]
```diff
@@ -463,62 +413,117 @@ def reviewer():
...
+                    "tracked_files_unchanged": clean,
+                    "current_head": current,
+                    "accepted": accepted,
+                },
+                indent=2,
+            )
+        )
+        # A transport/model failure with no complete report is retryable review
+        # work, not an instruction for the developer to change application code.
+        complete = (
```
[the remark this answers is not in the document]
Comment (neubig) [reply skills/github-software-factory/scripts/main.py:543]: Kept the retry-versus-revision distinction. In d46e943, interrupted review also posts a readable GitHub comment naming the failure category and retry behavior. Acceptance remains unset so infrastructure faults do not trigger application revisions.

File: skills/github-software-factory/scripts/extension_workflows.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        repository, issue, {"id": "ready-for-dev"}, branch, "main", base_sha
+    )
+    # Publication is a capability of the coordinator, not a workspace-kind choice.
+    return prompt + (
+        f"\nExecution contract: work in {workspace / 'project'}. "
+        f"Use `{workspace / 'bin/gh'} api` for the REST requests above; "
+        "it supplies a scoped credential. No GitHub token is needed. "
+        "The coordinator owns remote publication: leave your completed changes "
+        "and PR description in the workspace; do not push or create the PR yourself. "
```
[the remark this answers is not in the document]
Comment (neubig) [reply skills/github-software-factory/scripts/extension_workflows.py:78]: Fixed in d46e943: the canonical issue-to-PR builder accepts coordinator-owned publication and explicit GitHub access instructions. Factory prompts contain no direct git push, gh pr create, or personal-token instructions. A regression checks the composed prompt.

File: skills/github-software-factory/scripts/extension_workflows.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+def implementation_prompt(repository, issue, branch, base_sha, workspace, feedback):
+    workflow = module("skills/github-issue-to-pr/scripts/main.py")
+    prompt = workflow._build_implementation_prompt(
+        repository, issue, {"id": "ready-for-dev"}, branch, "main", base_sha
+    )
+    # Publication is a capability of the coordinator, not a workspace-kind choice.
+    return prompt + (
+        f"\nExecution contract: work in {workspace / 'project'}. "
+        f"Use `{workspace / 'bin/gh'} api` for the REST requests above; "
+        "it supplies a scoped credential. No GitHub token is needed. "
```
[the remark this answers is not in the document]
Comment (neubig) [reply skills/github-software-factory/scripts/extension_workflows.py:76]: Fixed in d46e943 through an explicit canonical builder mode. Factory prompts contain no direct publication commands or personal-token instructions; ordinary issue-to-PR retains its default publication behavior.

File: skills/github-software-factory/scripts/main.py [hunk 5]
```diff
@@ -422,38 +375,35 @@ def status(sha, context, passed, detail):
...
+    target = WORKSPACE / "bin" / "gh"
+    target.parent.mkdir(exist_ok=True)
+    shutil.copyfile(Path(__file__).with_name("scoped_gh.py"), target)
+    target.chmod(0o700)
+
+
+def gh_pages(endpoint):
+    items = []
+    for page in range(1, 101):
+        batch = gh("GET", f"{endpoint}?per_page=100&page={page}")
```
[the remark this answers is not in the document]
Comment (neubig) [reply skills/github-software-factory/scripts/main.py]: Fixed in d46e943: merge pagination parameters with urllib.parse so existing endpoint filters survive. Regression coverage exercises /pulls?state=open.

File: skills/github-software-factory/scripts/main.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (neubig) [reply skills/github-software-factory/scripts/main.py:543]: Fixed the silent failure in d46e943: an exception posts a readable comment identifying the failure category and retry behavior, and records it in the private acceptance artifact. This remains retryable reviewer work and cannot publish acceptance.

File: skills/github-software-factory/scripts/scoped_gh.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    parser = argparse.ArgumentParser(description=__doc__)
+    parser.add_argument("command", choices=["api"])
+    parser.add_argument("endpoint")
+    parser.add_argument("-X", "--method", default=None)
+    parser.add_argument("--input")
+    parser.add_argument("--paginate", action="store_true")
+    parser.add_argument("--jq", "-q")
+    parser.add_argument("-H", "--header", action="append", default=[])
+    args = parser.parse_args(argv)
+    config_path = config_path or Path(__file__).resolve().parents[1] / "config.json"
```
[the remark this answers is not in the document]
Comment (neubig) [reply skills/github-software-factory/scripts/scoped_gh.py:66]: Verified: automation #451 sets WORKSPACE_BASE to get_work_dir(run_id), the same directory used for extraction and execution. The identical-bundle local/Docker live probe validated this contract. The README now states it explicitly; the gh adapter and config.json share the bundle root in both modes.

**Comment by neubig** [comment]
Consolidating this canonical workflow composition directly into #557 during the requested PR audit. The initial custom JSON reviewer must not be merged first and immediately replaced. #557 now carries the final reusable issue-to-PR/reviewer/QA composition and retains the SDK release prerequisite. The new review order is #559 -> #557 -> #564. Closing this intermediate PR without merging it to main.

File: skills/github-software-factory/scripts/main.py [hunk 7]
```diff
@@ -336,10 +312,9 @@ def developer():
     if prs:
         for pr in prs:
             state = statuses(pr["head"]["sha"])
-            if any(
-                state.get(c) in ("failure", "error")
-                for c in ("software-factory/tests", "software-factory/review")
-            ):
+            # Wait for the independent review to publish its findings before
+            # revising, even if the earlier deterministic test phase failed.
+            if state.get("software-factory/review") in ("failure", "error"):
```
[the remark this answers is not in the document]
Comment (neubig) [reply skills/github-software-factory/scripts/main.py:318]: Intentional: a developer waits for completed review findings, while the reviewer retries incomplete infrastructure/model failures. This composition is now consolidated directly into #557; #562 is closed.

File: skills/github-software-factory/scripts/main.py [hunk 8]
```diff
@@ -478,62 +433,126 @@ def reviewer():
...
+                    "tracked_files_unchanged": clean,
+                    "current_head": current,
+                    "accepted": accepted,
+                },
+                indent=2,
+            )
+        )
+        # A transport/model failure with no complete report is retryable review
+        # work, not an instruction for the developer to change application code.
+        complete = (
```
[the remark this answers is not in the document]
Comment (neubig) [reply skills/github-software-factory/scripts/main.py:543]: The existing adjacent comment distinguishes retryable infrastructure failures from completed review outcomes. The consolidated #557 additionally avoids posting a code-revision status when final verification itself fails.

[the remark this answers is not in the document]
Comment (neubig) [reply skills/github-software-factory/scripts/main.py:543]: Retry failures already publish a visible PR comment with the conversation ID. Each retry is an independent exact-head review; previous reports do not silently satisfy a new run. This favors explicit evidence over reusing an unverified cross-run report. The final composition now lives in #557.

File: skills/github-software-factory/scripts/main.py [hunk 9]
```diff
@@ -384,42 +360,22 @@ def developer():
...
-    feedback = (
-        gh("GET", f"/issues/{existing['number']}/comments?per_page=100")
-        if existing
-        else []
-    )
-    issue["triage"] = gh("GET", f"/issues/{issue['number']}/comments?per_page=100")
+    feedback = {}
+    if existing:
+        for name, endpoint in (
+            ("discussion", f"/issues/{existing['number']}/comments"),
```
[the remark this answers is not in the document]
Comment (neubig) [reply skills/github-software-factory/scripts/main.py:366]: Fixed in a24a1e6 in consolidated #557: guard final verification, preserve original failures, and always persist fail-closed acceptance evidence. A regression test covers simultaneous review, gateway, and Git failures. #562 is superseded.
